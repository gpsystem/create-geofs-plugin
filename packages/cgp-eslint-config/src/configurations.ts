import type { EslintConfig } from "eslint-define-config";

export type EslintConfigNames = "baseConfig" | "reactBase" | "tsBase";

/**
 * The eslint configurations. All the values should be objects.
 */
export const eslintConfigBases: {
  [key in EslintConfigNames]: EslintConfig;
} = {
  baseConfig: {
    env: { browser: true, es2021: true, node: true },
    extends: ["eslint:recommended", "plugin:import/recommended"],
    globals: {
      GlassPanel: "readonly",
      Indicator: "readonly",
      Object3D: "readonly",
      Overlay: "readonly",
      PAGE_PATH: "readonly",
      PID: "readonly",
      absMin: "readonly",
      audio: "readonly",
      boundHours24: "readonly",
      clamp: "readonly",
      clone: "readonly",
      controls: "readonly",
      exponentialSmoothing: "readonly",
      exponentialSmoothingV3: "readonly",
      fixAngle: "readonly",
      fixAngle360: "readonly",
      fixAngles: "readonly",
      fixAngles360: "readonly",
      flight: "readonly",
      geoDecodeLocation: "readonly",
      geofs: "readonly",
      getBuildingCollision: "readonly",
      getURLParameters: "readonly",
      instruments: "readonly",
      ll2xy: "readonly",
      lla2xyz: "readonly",
      lookAt: "readonly",
      multiplayer: "readonly",
      objects: "readonly",
      rigidBody: "readonly",
      span: "readonly",
      ui: "readonly",
      weather: "readonly",
      xy2ll: "readonly",
      xyz2lla: "readonly",
    },
    parserOptions: { ecmaVersion: "latest", sourceType: "module" },
    root: true,
    rules: {
      eqeqeq: ["error"],
      "import/order": [
        "error",
        { alphabetize: { caseInsensitive: true, order: "asc" } },
      ],
    },
  },
  reactBase: {
    extends: ["plugin:react/recommended"],
    parserOptions: {
      ecmaFeatures: { jsx: true },
    },
    plugins: ["react"],
    settings: { react: { version: "detect" } },
  },
  tsBase: {
    extends: [
      "plugin:@typescript-eslint/recommended",
      "plugin:import/typescript",
    ],
    overrides: [
      {
        files: ["**/*.js"],
        rules: { "@typescript-eslint/no-var-requires": ["off"] },
      },
    ],
    parser: "@typescript-eslint/parser",
    plugins: ["@typescript-eslint"],
    settings: {
      "import/parsers": { "@typescript-eslint/parser": [".ts", ".tsx"] },
    },
  },
};

/**
 * The dependencies of the configurations. Indexable by the same keys that {@link eslintConfigBases} is by.
 */
export const eslintConfigBaseDependencies: Record<EslintConfigNames, string[]> =
  {
    baseConfig: ["eslint@v8", "eslint-plugin-import@v2"],
    reactBase: ["eslint-plugin-react@v7"],
    tsBase: [
      "@typescript-eslint/eslint-plugin@v5",
      "eslint-import-resolver-typescript@v2",
    ],
  };
