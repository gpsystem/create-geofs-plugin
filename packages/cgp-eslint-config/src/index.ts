import merge from "lodash.merge";
import {
  eslintConfigBaseDependencies,
  eslintConfigBases,
  EslintConfigNames,
  UnknownEslintConfigBase,
} from "./configurations";

export type MoreThanOneArray<T> = [T, ...T[]];

// TODO: is there a type for a eslint configuration file?
/**
 * Merges eslint configuration templates into one configuration to output to a .eslintrc file.
 *
 * @param configNames All the configuration templates.
 * The templates will be merged (deeply) with further right configuration templates overwriting further left templates.
 *
 * @returns A tuple. First is the configuration, ready to be stringified and outputted to a .eslintrc file.
 * Second is the names of the dependencies to install to make sure the .eslintrc configuration works properly.
 */
export function getEslintConfig(
  ...configNames: MoreThanOneArray<EslintConfigNames>
): [config: Record<string, unknown>, dependencies: string[]] {
  if (!configNames.length)
    throw new Error("must provide at least one config name");
  const configsToMerge: UnknownEslintConfigBase[] = configNames.map(
    (val) => eslintConfigBases[val]
  );
  const config: Record<string, unknown> = merge(
    ...(configsToMerge as MoreThanOneArray<UnknownEslintConfigBase>)
  );

  // if the typescript template is mentioned, remove globals from the config
  if (configNames.includes("tsBase") && config.globals !== undefined)
    delete config.globals;

  return [
    config,
    configNames.reduce<string[]>(
      (acc, val) => [...acc, ...eslintConfigBaseDependencies[val]],
      []
    ),
  ];
}

export { EslintConfigNames };
