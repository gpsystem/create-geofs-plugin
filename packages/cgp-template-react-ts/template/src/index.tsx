/// <reference types="jquery" />
/// <reference types="@geps/geofs-types" />

import React from "react";
import { createRoot } from "react-dom/client";
import { App } from "./App";

// Replace the selector with a link to your UI host!
// Example: #geofs-ui-right
// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
const root = createRoot(document.querySelector("REPLACE ME")!);
root.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>
);
