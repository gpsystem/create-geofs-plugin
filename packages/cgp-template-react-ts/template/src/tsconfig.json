{
  "compilerOptions": {
    "target": "es2015",
    "module": "esnext",
    "esModuleInterop": true,
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "jsx": "react",
    "skipLibCheck": true
  },
  "include": ["**/*"]
}
