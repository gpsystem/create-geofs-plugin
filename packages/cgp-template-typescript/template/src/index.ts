/// <reference types="jquery" />
/// <reference types="@geps/geofs-types" />

// start writing your code here!
// you can reference the geofs apis in intellisense because of the triple-slash directives
// the apis from geofs plugin system are not available yet
