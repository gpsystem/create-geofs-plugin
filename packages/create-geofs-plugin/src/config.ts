import { resolve } from "node:path";
import { cwd } from "node:process";
import { Command } from "commander";
import { sync as packageDirectorySync } from "pkg-dir";
import { version } from "../package.json";

export interface Configuration {
  targetDir: string;
  templateName: "default" | "typescript" | "react" | "react-ts";
  templateDir: string;
  overwrite: boolean;
}

export function parseConfig(argv: string[]): Configuration {
  let destination = "";

  // eslint-disable-next-line prefer-const
  let { overwrite, template } = new Command()
    .name("create-geofs-plugin")
    .version(`Create GeoFS Plugin v${version}`)
    .argument("<destination>")
    .action((dest: string) => {
      if (!dest.includes("/"))
        throw new Error("argument <destination> must be a relative path");
      else destination = resolve(cwd(), dest);
    })
    .option("--overwrite", "Overwrite existing files", false)
    .option("-t, --template <template>", "Name of the template to use")
    .parse(argv)
    .opts<{ overwrite?: boolean; template?: string }>();

  if (!["default", "typescript", "react", "react-ts"].includes(template ?? ""))
    template = "default";

  // TODO: create templates
  const templateDir: string | undefined = require.resolve(
    // `@geps/cgp-template-${template}`
    `@geps/cgp-eslint-config`
  );
  if (!templateDir) throw new Error("template not found");

  return {
    targetDir: destination,
    overwrite: overwrite ?? false,
    templateName: template as "default" | "typescript" | "react" | "react-ts",
    templateDir,
  };
}
