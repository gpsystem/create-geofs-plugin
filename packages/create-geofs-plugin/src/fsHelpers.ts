import {
  Dirent,
  existsSync,
  mkdirSync,
  readFileSync,
  readdirSync,
  statSync,
  writeFileSync,
} from "node:fs";
import { dirname, join, relative, resolve } from "node:path";

function pathIsDirectory(path: string): boolean {
  return statSync(path).isDirectory();
}

function internalWriteFileSync(targetFile: string, data: string): void {
  writeFileSync(targetFile, data, "utf-8");
}

export function outputFile(
  targetFile: string,
  data: string,
  force = false
): void {
  if (!force && existsSync(targetFile)) throw new Error("file already exists");

  const targetFileDirectory: string = dirname(targetFile);

  if (!existsSync(targetFileDirectory))
    mkdirSync(targetFileDirectory, { recursive: true });
  internalWriteFileSync(targetFile, data);
}

// TODO: should we make a wrapper for this that throws on initialization and returns the full string[]?
export function* getAllFilesInDir(
  dirName: string
): Generator<string, void, void> {
  if (!pathIsDirectory(dirName))
    throw new Error(`passed path is not a directory`);

  const directoryDetails: Dirent[] = readdirSync(dirName, {
    withFileTypes: true,
  });

  for (const directoryChild of directoryDetails) {
    const childPath: string = resolve(dirName, directoryChild.name);

    if (directoryChild.isDirectory()) yield* getAllFilesInDir(childPath);
    else if (directoryChild.isFile()) yield childPath;
  }
}

export interface CopyDirOptions {
  overwrite?: boolean;
}

export function copyDir(
  srcDir: string,
  targetDir: string,
  { overwrite = false }: CopyDirOptions = {}
): void {
  if (srcDir === targetDir)
    throw new Error("source and target directories can't be the same");
  if (!pathIsDirectory(srcDir))
    throw new Error("source path isn't a directory");

  for (const srcFilePath of getAllFilesInDir(srcDir)) {
    const newFilePath: string = join(targetDir, relative(srcDir, srcFilePath));

    if (existsSync(newFilePath) && overwrite === false) continue;
    else outputFile(newFilePath, readFileSync(srcFilePath, "utf-8"), overwrite);
  }
}
