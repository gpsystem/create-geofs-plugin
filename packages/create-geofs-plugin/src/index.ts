import { join } from "node:path";
import { dir, DirectoryResult } from "tmp-promise";
import { Configuration, parseConfig } from "./config";
import { copyDir } from "./fsHelpers";
import expandTemplateJson, { ExpandedTemplateJson } from "./template";

export async function start(argv: string[]): Promise<void> {
  const config: Configuration = parseConfig(argv);
  const { path: tmpDirName, cleanup: cleanupTempDir }: DirectoryResult =
    await dir({
      unsafeCleanup: true,
    });

  copyDir(config.templateDir, tmpDirName, { overwrite: config.overwrite });
  const { dependencies: templateDeps, eslintConfig }: ExpandedTemplateJson =
    expandTemplateJson(join(tmpDirName, "template.json"));
  // remove everything that starts with __
  // copy everything to the target dir
  // install the dependencies
  await cleanupTempDir();
}
