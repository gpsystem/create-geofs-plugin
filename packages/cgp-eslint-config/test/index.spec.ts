import { describe, expect, test } from "@jest/globals";
import { eslintConfigBases } from "../src/configurations";
import { getEslintConfig } from "../src/index";

describe("get the eslint config", () => {
  const allConfigBasesNames: (keyof typeof eslintConfigBases)[] = [
    "baseConfig",
    "tsBase",
    "reactBase",
  ];
  for (const configBaseName of Object.keys(eslintConfigBases)) {
    if (
      !allConfigBasesNames.includes(
        configBaseName as keyof typeof eslintConfigBases
      )
    )
      throw new Error(
        "allConfigBasesNames needs to include all config base names before a test can be run"
      );
  }

  test("requires one config name", () => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore expecting an error
    expect(() => getEslintConfig()).toThrowErrorMatchingSnapshot();
    expect(() => getEslintConfig("baseConfig")).not.toThrow();
  });

  test("returns an array with the config and dependencies", () => {
    const [config, dependencies] = getEslintConfig("baseConfig");
    // config is an object
    expect(Object.prototype.toString.call(config)).toBe(
      Object.prototype.toString()
    );
    expect(Array.isArray(dependencies)).toBe(true);
  });

  test("returns a dependency string array", () => {
    const [, dependencies] = getEslintConfig("baseConfig");

    expect(Array.isArray(dependencies)).toBe(true);
    expect(dependencies.every((val) => typeof val === "string")).toBe(true);
  });

  test("returns consistent values", () => {
    /**
     * Rules for a permutation:
     * 1. Must have at least one element
     * 2. Must have baseConfig first
     */
    const validConfigBaseNamePermutations: [
      keyof typeof eslintConfigBases,
      ...(keyof typeof eslintConfigBases)[]
    ][] = [
      ["baseConfig"],
      ["baseConfig", "reactBase"],
      ["baseConfig", "tsBase"],
      ["baseConfig", "reactBase", "tsBase"],
      ["baseConfig", "tsBase", "reactBase"],
    ];
    for (const configBaseNames of validConfigBaseNamePermutations) {
      const [config, dependencies] = getEslintConfig(...configBaseNames);
      const runName: string = configBaseNames
        .map((val) => val.split(/[A-Z]/)[0])
        .join("-");

      expect(config).toMatchSnapshot(`config for ${runName} run`);
      expect(dependencies).toMatchSnapshot(`dependencies for ${runName} run`);
    }
  });
});
