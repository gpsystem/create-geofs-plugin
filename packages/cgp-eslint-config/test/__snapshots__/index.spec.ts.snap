// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get the eslint config requires one config name 1`] = `"must provide at least one config name"`;

exports[`get the eslint config returns consistent values: config for base run 1`] = `
Object {
  "env": Object {
    "browser": true,
    "es2021": true,
    "node": true,
  },
  "extends": Array [
    "eslint:recommended",
    "plugin:import/recommended",
  ],
  "globals": Object {
    "GlassPanel": "readonly",
    "Indicator": "readonly",
    "Object3D": "readonly",
    "Overlay": "readonly",
    "PAGE_PATH": "readonly",
    "PID": "readonly",
    "absMin": "readonly",
    "audio": "readonly",
    "boundHours24": "readonly",
    "clamp": "readonly",
    "clone": "readonly",
    "controls": "readonly",
    "exponentialSmoothing": "readonly",
    "exponentialSmoothingV3": "readonly",
    "fixAngle": "readonly",
    "fixAngle360": "readonly",
    "fixAngles": "readonly",
    "fixAngles360": "readonly",
    "flight": "readonly",
    "geoDecodeLocation": "readonly",
    "geofs": "readonly",
    "getBuildingCollision": "readonly",
    "getURLParameters": "readonly",
    "instruments": "readonly",
    "ll2xy": "readonly",
    "lla2xyz": "readonly",
    "lookAt": "readonly",
    "multiplayer": "readonly",
    "objects": "readonly",
    "rigidBody": "readonly",
    "span": "readonly",
    "ui": "readonly",
    "weather": "readonly",
    "xy2ll": "readonly",
    "xyz2lla": "readonly",
  },
  "parserOptions": Object {
    "ecmaVersion": "latest",
    "sourceType": "module",
  },
  "root": true,
  "rules": Object {
    "eqeqeq": Array [
      "error",
    ],
    "import/order": Array [
      "error",
      Object {
        "alphabetize": Object {
          "caseInsensitive": true,
          "order": "asc",
        },
      },
    ],
  },
}
`;

exports[`get the eslint config returns consistent values: config for base-react run 1`] = `
Object {
  "env": Object {
    "browser": true,
    "es2021": true,
    "node": true,
  },
  "extends": Array [
    "plugin:react/recommended",
    "plugin:import/recommended",
  ],
  "globals": Object {
    "GlassPanel": "readonly",
    "Indicator": "readonly",
    "Object3D": "readonly",
    "Overlay": "readonly",
    "PAGE_PATH": "readonly",
    "PID": "readonly",
    "absMin": "readonly",
    "audio": "readonly",
    "boundHours24": "readonly",
    "clamp": "readonly",
    "clone": "readonly",
    "controls": "readonly",
    "exponentialSmoothing": "readonly",
    "exponentialSmoothingV3": "readonly",
    "fixAngle": "readonly",
    "fixAngle360": "readonly",
    "fixAngles": "readonly",
    "fixAngles360": "readonly",
    "flight": "readonly",
    "geoDecodeLocation": "readonly",
    "geofs": "readonly",
    "getBuildingCollision": "readonly",
    "getURLParameters": "readonly",
    "instruments": "readonly",
    "ll2xy": "readonly",
    "lla2xyz": "readonly",
    "lookAt": "readonly",
    "multiplayer": "readonly",
    "objects": "readonly",
    "rigidBody": "readonly",
    "span": "readonly",
    "ui": "readonly",
    "weather": "readonly",
    "xy2ll": "readonly",
    "xyz2lla": "readonly",
  },
  "parserOptions": Object {
    "ecmaFeatures": Object {
      "jsx": true,
    },
    "ecmaVersion": "latest",
    "sourceType": "module",
  },
  "plugins": Array [
    "react",
  ],
  "root": true,
  "rules": Object {
    "eqeqeq": Array [
      "error",
    ],
    "import/order": Array [
      "error",
      Object {
        "alphabetize": Object {
          "caseInsensitive": true,
          "order": "asc",
        },
      },
    ],
  },
  "settings": Object {
    "react": Object {
      "version": "detect",
    },
  },
}
`;

exports[`get the eslint config returns consistent values: config for base-react-ts run 1`] = `
Object {
  "env": Object {
    "browser": true,
    "es2021": true,
    "node": true,
  },
  "extends": Array [
    "plugin:@typescript-eslint/recommended",
    "plugin:import/typescript",
  ],
  "overrides": Array [
    Object {
      "files": Array [
        "**/*.js",
      ],
      "rules": Object {
        "@typescript-eslint/no-var-requires": Array [
          "off",
        ],
      },
    },
  ],
  "parser": "@typescript-eslint/parser",
  "parserOptions": Object {
    "ecmaFeatures": Object {
      "jsx": true,
    },
    "ecmaVersion": "latest",
    "sourceType": "module",
  },
  "plugins": Array [
    "@typescript-eslint",
  ],
  "root": true,
  "rules": Object {
    "eqeqeq": Array [
      "error",
    ],
    "import/order": Array [
      "error",
      Object {
        "alphabetize": Object {
          "caseInsensitive": true,
          "order": "asc",
        },
      },
    ],
  },
  "settings": Object {
    "import/parsers": Object {
      "@typescript-eslint/parser": Array [
        ".ts",
        ".tsx",
      ],
    },
    "react": Object {
      "version": "detect",
    },
  },
}
`;

exports[`get the eslint config returns consistent values: config for base-ts run 1`] = `
Object {
  "env": Object {
    "browser": true,
    "es2021": true,
    "node": true,
  },
  "extends": Array [
    "plugin:@typescript-eslint/recommended",
    "plugin:import/typescript",
  ],
  "overrides": Array [
    Object {
      "files": Array [
        "**/*.js",
      ],
      "rules": Object {
        "@typescript-eslint/no-var-requires": Array [
          "off",
        ],
      },
    },
  ],
  "parser": "@typescript-eslint/parser",
  "parserOptions": Object {
    "ecmaFeatures": Object {
      "jsx": true,
    },
    "ecmaVersion": "latest",
    "sourceType": "module",
  },
  "plugins": Array [
    "@typescript-eslint",
  ],
  "root": true,
  "rules": Object {
    "eqeqeq": Array [
      "error",
    ],
    "import/order": Array [
      "error",
      Object {
        "alphabetize": Object {
          "caseInsensitive": true,
          "order": "asc",
        },
      },
    ],
  },
  "settings": Object {
    "import/parsers": Object {
      "@typescript-eslint/parser": Array [
        ".ts",
        ".tsx",
      ],
    },
    "react": Object {
      "version": "detect",
    },
  },
}
`;

exports[`get the eslint config returns consistent values: config for base-ts-react run 1`] = `
Object {
  "env": Object {
    "browser": true,
    "es2021": true,
    "node": true,
  },
  "extends": Array [
    "plugin:react/recommended",
    "plugin:import/typescript",
  ],
  "overrides": Array [
    Object {
      "files": Array [
        "**/*.js",
      ],
      "rules": Object {
        "@typescript-eslint/no-var-requires": Array [
          "off",
        ],
      },
    },
  ],
  "parser": "@typescript-eslint/parser",
  "parserOptions": Object {
    "ecmaFeatures": Object {
      "jsx": true,
    },
    "ecmaVersion": "latest",
    "sourceType": "module",
  },
  "plugins": Array [
    "react",
  ],
  "root": true,
  "rules": Object {
    "eqeqeq": Array [
      "error",
    ],
    "import/order": Array [
      "error",
      Object {
        "alphabetize": Object {
          "caseInsensitive": true,
          "order": "asc",
        },
      },
    ],
  },
  "settings": Object {
    "import/parsers": Object {
      "@typescript-eslint/parser": Array [
        ".ts",
        ".tsx",
      ],
    },
    "react": Object {
      "version": "detect",
    },
  },
}
`;

exports[`get the eslint config returns consistent values: dependencies for base run 1`] = `
Array [
  "eslint@v8",
  "eslint-plugin-import@v2",
]
`;

exports[`get the eslint config returns consistent values: dependencies for base-react run 1`] = `
Array [
  "eslint@v8",
  "eslint-plugin-import@v2",
  "eslint-plugin-react@v7",
]
`;

exports[`get the eslint config returns consistent values: dependencies for base-react-ts run 1`] = `
Array [
  "eslint@v8",
  "eslint-plugin-import@v2",
  "eslint-plugin-react@v7",
  "@typescript-eslint/eslint-plugin@v5",
  "eslint-import-resolver-typescript@v2",
]
`;

exports[`get the eslint config returns consistent values: dependencies for base-ts run 1`] = `
Array [
  "eslint@v8",
  "eslint-plugin-import@v2",
  "@typescript-eslint/eslint-plugin@v5",
  "eslint-import-resolver-typescript@v2",
]
`;

exports[`get the eslint config returns consistent values: dependencies for base-ts-react run 1`] = `
Array [
  "eslint@v8",
  "eslint-plugin-import@v2",
  "@typescript-eslint/eslint-plugin@v5",
  "eslint-import-resolver-typescript@v2",
  "eslint-plugin-react@v7",
]
`;
